---
title: "Detecting and Simulating Punctuated Evolution"
author: "Kevin Surya"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Detecting and Simulating Punctuated Evolution}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global_options, include = FALSE}
# Set global options ----
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE,
  dev = "svg",
  fig.align = "center"
)
options(mc.cores = parallel::detectCores())
```

```{css formatting, echo = FALSE}
/* Prevent superscripts from increasing line heights */
/* Codes adapted from https://stackoverflow.com/questions/1530685/html-sup-tag-affecting-line-height-how-to-make-it-consistent and https://bookdown.org/yihui/rmarkdown-cookbook/html-css.html*/
sup {
  line-height: 0;
  font-size: 0.6em;
  vertical-align: super;
}

/* Set the color of hyperlinks */
a {
  color: black;
}

/* Remove border box around HTML outputs */
/* Codes adapted from https://stackoverflow.com/questions/64828727/how-to-remove-border-box-around-r-markdown-html-output-using-knitr-chunk-options */
img {
  border: 0;
}
```

## Why should we care about punctuated evolution?

Explaining the diversity of life today and in the past is the grail of evolutionary biology. Darwin^1^ predicted that variation emerges gradually, where evolutionary change accumulates at a steady rate over long timespans. However, the fossil record suggests that new traits often appear abruptly, coincident with events where a subpopulation splits off from the main population^2,3^. We call these events speciations. The increased evolutionary rate during speciation could be due to drift related to small population size^2^ or adaptive as a subpopulation enters a new environment^3^. Eldredge and Gould^3^ noticed this pattern and posited that a lineage's history comprises long periods of stasis punctuated by rapid evolution around speciation events.

But, the idea that ~100% of divergence happens during speciation is extreme. Rather than debating this view of evolution (gradual vs. punctuated), we can estimate how much the variability of a trait in a group is attributable to speciation events (0% to 100%). Mark Pagel and colleagues at the University of Reading devised such methods (which we will review below) to infer that net speciation explains ~16% of the total molecular evolution in extant plants, fungi, and animals^4,5^. The method applies to any evolving system, including human languages^6^. Up to 33% of vocabulary differences are attributable to language-splitting events^6^. Punctuated evolution is, therefore, a widespread pattern in how genomes, phenotypes, and cultures evolve. And we need to quantify this pattern.

Punctuated evolution alone does not explain the diversity of life, and it is not the only one proposed. For a brief overview of current hypotheses or models, I suggest reading Pagel et al.^7^. We should not treat the method for detecting punctuated evolution as a 'hammer' to resolve every problem. But, we can test the extent to which the pattern observable in a dataset is consistent with punctuated evolution.

## Why `fallpaddy`?

There are a few reasons for developing the `fallpaddy` package. My colleagues and I are generalizing Pagel et al.'s^5^ method to cover serially sampled taxa, such as extinct dinosaurs and viruses, including SARS-CoV-2 (Surya et al., in review). I want this modified model to be easily accessible to evolutionary biologists, including myself. No one wants to keep writing long `R` or `Python` scripts if they can streamline the process with functions. Second, this package has a function for simulating punctuated evolution in a manner more flexible than in other packages. This function allows the user to specify the punctuational contribution to the total trait divergence (0 to 1) instead of fixing it to one. Lastly, this package is a one-stop shop for studying punctuated evolution. It contains tools for data extraction, wrangling, visualization, and diagnostics.

This package, however, uses maximum likelihood, not Bayesian. `fallpaddy` is most useful for quick-and-dirty exploratory analyses or if the dataset size precludes the computationally heavy Bayesian workhorse, Markov chain Monte Carlo (MCMC). My colleagues and I will not be developing Bayesian functions in this package. Expect them to be available in the `Python` library `Sci-Phy` [(link)](https://chrisorgan.github.io/code/) in the future.

This vignette document showcases how to use `fallpaddy` primary functions in four case studies, covering molecular and phenotypic traits and distant taxonomic groups, from dinosaurs to viruses. So, let's begin by loading the package. I already bundled the datasets with the package.

```{r load_package}
# Load package ----
library(fallpaddy)
```

## 1. Molecular evolution in snakes and lizards

Here, we will apply the standard Pagel et al.'s^5^ model to a molecular, extant-only dataset. The dataset is a molecular phylogenetic tree of Lepidosauria (tuatara, snakes, and lizards) comprising 4,162 species. I took the tree from Pyron et al.^8^. They inferred a maximum likelihood tree from a concatenated alignment of five mitochondrial and seven nuclear genes (12,896 bp in total). It is rooted and fully bifurcating. Below is a plot of the tree. Ideally, I would color the branches by clades and delete the axis line and axis ticks using `Adobe Illustrator` to maximize the data-ink ratio^9^. But, for a quick tree visualization, this plot is sufficient.

```{r 1_plot_tree, fig.cap="Fig. 1: Molecular phylogenetic tree of 4,162 lepidosaur species from Pyron et al.^8^."}
# Plot tree ----
plot_tree(tree = tree_lepidosaur_mol, unit = "subs/site")
```

This tree contains the information we need to detect punctuated evolution. The first task is to estimate the predicted difference in net molecular evolution between two lepidosaur species where one of them has undergone one additional net-speciation event^5^. A species' net-molecular divergence is proxied by its phylogenetic path length or root-to-tip distance. A net-speciation event is the cumulative number of branching events (i.e., node count) along a species' root-to-tip path. Net speciation is speciation minus extinction. So, we need to extract the path length and node count from the tree.

```{r 1_extract_data}
# Extract path length and node count ----
path1 <- get_path_length(tree = tree_lepidosaur_mol)
node1 <- get_node_count(tree = tree_lepidosaur_mol)

# Prepare the data frame ----
data1 <- data.frame(path = path1, node = node1)
# Reorder rows in the data frame to follow the order of tree tip labels
data1 <- reorder_data(data = data1, tree = tree_lepidosaur_mol)
head(data1)
```

It is always good to plot the data before the analyses^10^. Benefits include familiarizing ourselves with the data structure, which covers variable distributions, correlations, and outliers or unexpected data points^10^. Data visualization indicates how the data would change our a priori beliefs.

```{r 1_plot_data, fig.cap = "Fig. 2: Scatter plot of the net molecular divergence (phylogenetic path length) vs. net-speciation event (node count) in Lepidosauria."}
# Plot data ----
plot_data_2d(
  data = data1,
  unit = "subs/site",
  pt_size = 1,
  alpha = 0.15,
  plot_width = 8.33,  # width of the HTML document
  text_size = 14,
  tick_size = 12
)
```

<br/>

The plot above is interactive. You can hover your cursor over a data point to see its information (species, path length, and node count). The data point in the lower left corner, the tuatara (*Sphenodon punctatus*), a lizard-like creature from New Zealand, is likely an influential outlier. Its path length and node count are much lower than in other species. Ideally, I would run the analyses with and without tuatara to assess its influence on estimated parameter values. The distributions of the individual variables are difficult to discern from the plot, given the large sample size. Plotting their histograms may be necessary. Otherwise, the scatter plot indicates a weak correlation between path length and node count.

Next, we will fit regression models. The following two steps are necessary to run the models, but you need not fully understand them.

```{r 1_prepare_matrices}
# Decompose the phylogenetic variance-covariance matrix into its correlation
#   (off-diagonal) and variance components (on-diagonal) ----
vcv1 <- decomp_vcv(tree = tree_lepidosaur_mol, lambda = 1)

# Create the normalizing matrix D ----
dmat1 <- create_dmat(tree = tree_lepidosaur_mol)
```

We can encode a phylogenetic tree as a variance-covariance matrix^11^, where each cell represents the shared amount of evolutionary history for a pair of species (covariance). Each cell on the matrix diagonals is the path length (variance). In general, this matrix is how we account for the non-independence between species^12^ and unequal variance (differing path lengths). Simple linear regression models assume equal variance and independence of the residuals or data points. The function `decomp_vcv` decomposes the matrix into its off-diagonal and on-diagonal components necessary for the downstream function `fit_punc_model`. The `create_dmat` function generates the normalizing matrix $D$, which contributes to decorrelating the data. For more details regarding the modifications of a simple regression model to accommodate phylogenetic data, read Symonds and Blomberg^13^.

## 2. Body size evolution in mammals

## 3. Genomic evolution in Zika viruses

## 4. Body size evolution in Mesozoic dinosaurs

## References

1. Darwin, C. *On the Origin of Species by Means of Natural Selection, or the Preservation of Favoured Races in the Struggle for Life*. (John Murray, 1859).

2. Mayr, E. Change of genetic environment and evolution. in *Evolution as a Process* (eds. Huxley, J., Hardy, A. C. & Ford, E. B.) 157–180 (Allan & Unwin, 1954).

3. Eldredge, N. & Gould, S. J. Punctuated equilibria: An alternative to phyletic gradualism. in *Models in Paleobiology* (ed. Schopf, T. J. M.) 82–115 (Freeman, Cooper, 1972).

4. Webster, A. J., Payne, R. J. H. & Pagel, M. Molecular phylogenies link rates of evolution and speciation. *Science* __301__, 478–478 (2003).

5. Pagel, M., Venditti, C. & Meade, A. Large punctuational contribution of speciation to evolutionary divergence at the molecular level. *Science* __314__, 119–121 (2006).

6. Atkinson, Q. D., Meade, A., Venditti, C., Greenhill, S. J. & Pagel, M. Languages evolve in punctuational bursts. *Science* __319__, 588–588 (2008).

7. Pagel, M., O'Donovan, C. & Meade, A. General statistical model shows that macroevolutionary patterns and processes are consistent with Darwinian gradualism. *Nat. Commun.* __13__, 1113 (2022).

8. Pyron, R. A., Burbrink, F. T. & Wiens, J. J. A phylogeny and revised classification of Squamata, including 4161 species of lizards and snakes. *BMC Evol. Biol.* __13__, 93 (2013).

9. Tufte, E. R. *The Visual Display of Quantitative Information*. (Graphics Press, 1983).

10. Steel, E. A., Kennedy, M. C., Cunningham, P. G. & Stanovick, J. S. Applied statistics in ecology: Common pitfalls and simple solutions. *Ecosphere* __4__, art115 (2013).

11. Grafen, A. The phylogenetic regression. *Philos. Trans. R. Soc. Lond. B Biol. Sci.* __326__, 119–157 (1989).

12. Felsenstein, J. Phylogenies and the comparative method. *Am. Nat.* __125__, 1–15 (1985).

13. Symonds, M. R. E. & Blomberg, S. P. A primer on phylogenetic generalised least squares. in *Modern Phylogenetic Comparative Methods and Their Application in Evolutionary Biology: Concepts and Practice* (ed. Garamszegi, L. Z.) 105–130 (Springer Berlin Heidelberg, 2014).
