---
title: "Detecting and Simulating Punctuated Evolution"
author: "Kevin Surya"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Detecting and Simulating Punctuated Evolution}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global_options, include = FALSE}
# Set global options ----
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE,
  dev = "svg",
  fig.align = "center"
)
options(mc.cores = parallel::detectCores())
```

```{css formatting, echo = FALSE}
/* Prevent superscripts from increasing line heights */
/* Codes adapted from https://stackoverflow.com/questions/1530685/html-sup-tag-affecting-line-height-how-to-make-it-consistent and https://bookdown.org/yihui/rmarkdown-cookbook/html-css.html*/
sup {
  line-height: 0;
  font-size: 0.6em;
  vertical-align: super;
}

/* Set the color of hyperlinks */
a {
  color: black;
}

/* Remove border box around HTML outputs */
/* Codes adapted from https://stackoverflow.com/questions/64828727/how-to-remove-border-box-around-r-markdown-html-output-using-knitr-chunk-options */
img {
  border: 0;
}
```

## Why should we care about punctuated evolution?

Explaining the diversity of life today and in the past is the grail of evolutionary biology. Darwin^1^ predicted that variation emerges gradually, where evolutionary change accumulates incrementally over long timespans. However, the fossil record suggests that new traits often appear abruptly, coincident with events where a subpopulation splits from the main population^2,3^. We call these events speciations. The increased evolutionary rate during speciation could be due to drift related to small population size^2^ or adaptive as a subpopulation enters a new environment^3^. Eldredge and Gould^3^ noticed this pattern and hypothesized that a lineage's history comprises long periods of stasis punctuated by rapid evolution around speciation events.

But, the idea that ~100% of divergence happens during speciation is extreme. Rather than debating gradual versus punctuated scenarios, we can estimate how much the variability of a trait in a group is attributable to speciation events (0%–100%). Mark Pagel and colleagues at the University of Reading devised such methods (which we will review below) to infer that net speciation explains ~16% of the total molecular evolution in extant plants, fungi, and animals^4,5^. The method applies to any evolving system, including human languages^6^. Up to 33% of vocabulary differences are attributable to language-splitting events^6^. Punctuated evolution is, therefore, a widespread pattern in how genomes, phenotypes, and cultures evolve. We need to quantify this pattern.

Punctuated evolution alone does not explain the diversity of life, and it is not the only one proposed. For a brief overview of current hypotheses or models, I suggest reading Pagel et al.^7^. We should not treat the method for detecting punctuated evolution as a 'hammer' to resolve every problem. But, we can test the extent to which the pattern observable in a dataset is consistent with punctuated evolution.

## Why `fallpaddy`?

There are a few reasons for developing the `fallpaddy` package. My colleagues and I are generalizing Pagel et al.'s^5^ method to cover serially sampled taxa, such as extinct dinosaurs and viruses, including SARS-CoV-2^8^. I want this modified model to be easily accessible to evolutionary biologists, including myself. No one wants to keep writing long `R` or `Python` scripts if they can streamline the process with functions. Second, this package has a function for simulating punctuated evolution in a manner more flexible than in other packages. This function allows the user to specify the punctuational contribution to the total trait divergence (0 to 1) instead of fixing it to one. Lastly, this package is a one-stop shop for studying punctuated evolution. It contains tools for data extraction, wrangling, visualization, and diagnostics.

This package, however, uses maximum likelihood, not Bayesian. `fallpaddy` is most useful for quick exploratory analyses or if the dataset size precludes the computationally heavy Bayesian workhorse, Markov chain Monte Carlo (MCMC). My colleagues and I will not be developing Bayesian functions in this package. Expect their availability in the `Python` library `Sci-Phy` [(link)](https://chrisorgan.github.io/code/) in the future.

This vignette document showcases how to use `fallpaddy` main functions in four case studies, covering molecular and phenotypic traits and distant taxonomic groups, from dinosaurs to viruses. So, let's begin by loading the package. I already bundled the datasets with the package.

```{r load_package}
# Load package ----
library(fallpaddy)
```

## 1. Molecular evolution in snakes and lizards

Here, we will apply the standard Pagel et al.'s^5^ model to a molecular, extant-only dataset. The dataset is a molecular phylogenetic tree of Lepidosauria (tuatara, snakes, and lizards) comprising 4,162 species. I took the tree from Pyron et al.^9^. They inferred a maximum likelihood tree from a concatenated alignment of five mitochondrial and seven nuclear genes (12,896 bp in total). It is rooted and fully bifurcating. Below is a plot of the tree. Ideally, I would color the branches by clades and delete the axis line and axis ticks using Adobe Illustrator to maximize the data-ink ratio^10^. But, for a quick tree visualization, this plot is sufficient.

```{r 1_plot_tree, fig.cap = "Fig. 1: Molecular phylogenetic tree of 4,162 lepidosaur species from Pyron et al.^9^."}
# Plot tree ----
plot_tree(
  tree = tree_lepidosaur_mol,
  unit = "expected number of DNA substitutions per site"
)
# If you encounter an error, try updating all your `R` packages
```

This tree contains the information we need to detect punctuated evolution. The first task is to estimate the predicted difference in net molecular evolution between two lepidosaur species where one of them has undergone one additional net-speciation event^5^. A species' net-molecular divergence is proxied by its phylogenetic path length or root-to-tip distance. A net-speciation event is the cumulative number of branching events (i.e., node count) along a species' root-to-tip path. Net speciation is speciation minus extinction. So, we need to extract the path length and node count from the tree. See Pagel et al.'s^5^ Figure 1 for a simple schematic.

```{r 1_extract_data}
# Extract path length and node count ----
path1 <- get_path_length(tree = tree_lepidosaur_mol)
node1 <- get_node_count(tree = tree_lepidosaur_mol)

# Prepare the data frame ----
data1 <- data.frame(path = path1, node = as.integer(node1))
# Reorder rows in the data frame to follow the order of tree tip labels
data1 <- reorder_data(data = data1, tree = tree_lepidosaur_mol)

# Preview the dataset ----
head(data1)
```

It is always good to plot the data before the analyses^11^. Benefits include familiarizing with the data structure, which covers variable distributions, correlations, and outliers or unexpected data points^11^. Data visualization indicates how the data would change our a priori beliefs.

```{r 1_plot_data, fig.cap = "Fig. 2: Scatter plot of the net molecular divergence (phylogenetic path length) vs. net-speciation event (node count) in Lepidosauria."}
# Plot data ----
plot_data_2d(
  data = data1,
  unit = "DNA substitutions/site",
  pt_size = 1,
  alpha = 0.15,
  plot_width = 8.33,  # width of the HTML document
  text_size = 14,
  tick_size = 12
)
```

<br/>

The plot above is interactive. You can hover your cursor over a data point to see its species name, path length, and node count. The data point in the lower left corner, the tuatara (*Sphenodon punctatus*), a lizard-like creature from New Zealand, is most likely an outlier. Its path length and node count are much lower than in other species. Ideally, I would run the analyses with and without tuatara to assess its influence on estimated parameter values. However, for this document's brevity, I remove tuatara from the dataset (see the following code chunk). Hereafter, the scope of inference becomes squamates (lizards and snakes) instead of lepidosaurs. The distributions of the individual variables are difficult to discern from the plot, given the large sample size. Plotting their histograms is one workaround. Otherwise, the scatter plot indicates a weak correlation between path length and node count.

```{r 1_remove_tuatara}
# Remove tuatara from the phylogenetic tree ----
tree_squamate_mol <- ape::drop.tip(
  phy = tree_lepidosaur_mol,
  tip = "Sphenodon_punctatus"
)

# Re-assemble the data frame ----
path1 <- get_path_length(tree = tree_squamate_mol)
node1 <- get_node_count(tree = tree_squamate_mol)
data1 <- data.frame(path = path1, node = as.integer(node1))
data1 <- reorder_data(data = data1, tree = tree_squamate_mol)
```

Next, we will fit regression models. The following two steps are necessary to run the models, but you need not fully understand them.

```{r 1_prepare_matrices, eval = FALSE}
# Decompose the phylogenetic variance-covariance matrix into its correlation
#   (off-diagonal) and variance components (on-diagonal) ----
vcv1 <- decomp_vcv(tree = tree_squamate_mol, lambda = 1)

# Create the normalizing matrix D ----
dmat1 <- create_dmat(tree = tree_squamate_mol)
```

```{r 1_prepare_matrices_save_objects, eval = FALSE, echo = FALSE}
# To decrease knitting time, I save the created objects above ----
save(vcv1, file = "vcv1.RData")
save(dmat1, file = "dmat1.RData")
```

```{r 1_prepare_matrices_load_objects, echo = FALSE}
# To decrease knitting time, I load the created objects above ----
load("vcv1.RData")  # vcv1
load("dmat1.RData")  # dmat1
```

We can encode a phylogenetic tree as a variance-covariance matrix^12^, where each cell represents the shared amount of evolutionary history for a pair of species (covariance). Each cell on the matrix's diagonals is the path length (variance). In general, this matrix is how we account for species non-independence^13^ and unequal variance (differing path lengths). The function `decomp_vcv` decomposes the matrix into its off-diagonal and on-diagonal components necessary for the downstream function `fit_punc_model`. The `create_dmat` function generates the normalizing matrix $D$, which contributes to decorrelating the data. For more details regarding the generalization of a simple regression model to accommodate phylogenetic data, read Symonds and Blomberg^14^.

Next, we fit a null, baseline model, regressing path length on itself (`path ~ 1` or $path = \beta_0 + \epsilon$, where $\beta_0$ is the state at the root of the tree [intercept], and $\epsilon$ is the residual error). This mean-only model is the simplest possible model. The estimates are the phylogenetically-normalized mean and variance of the path lengths (not the raw mean and variance). Another measure calculated here is the sum of squared total ($SST$), crucial for measuring how much node count explains path length in the next model.

```{r 1_fit_null_model, eval = FALSE}
# Fit the null regression model ----
null_model1 <- fit_punc_model(
  data = data1,
  vcv_parts = vcv1,
  D = dmat1,
  model = "p"  # path ~ 1
)
```

```{r 1_fit_null_model_save_object, eval = FALSE, echo = FALSE}
# To decrease knitting time, I save the created object above ----
save(null_model1, file = "null_model1.RData")
```

```{r 1_fit_null_model_load_object, echo = FALSE}
# To decrease knitting time, I load the created object above ----
load("null_model1.RData")  # null_model1
```

```{r 1_show_null_model_outputs}
# Print the model outputs ----
summary(null_model1$model)
null_model1$sig2  # variance
null_model1$sse  # SST (or sum of squared estimates of error [SSE] of the null)
```

The estimated mean path length or the net molecular divergence in Squamata is $0.578 \pm 0.045$ DNA substitutions/site ($\bar{x} \pm 1.96 \times SEM$). We can change the unit to enhance interpretability (e.g., $578 \pm 45$ substitutions/1,000 sites). The estimated variance is $0.00698$ substitutions/site (or $6.98$ substitutions/1,000 sites). Taking the square root of the variance gives us the estimated standard deviation ($\hat{\sigma} = 0.0835$). This value means that the "average" difference between a path length and the expected value is $83.5$ substitutions/1,000 sites. Another way to view these numbers is that they are the expected mean and variance, assuming a strict molecular clock^15^. This clock model posits that substitutions accumulate gradually at a constant rate through time and across lineages.

Now, we fit the alternative model, which assesses the degree of the punctuational effect. Punctuated evolution means that substitutions tend to accumulate during speciation events. If true, the evolutionary rate would not be clock-like. This model estimates the extent that node count drives deviations from the strict molecular clock (`path ~ node` or $path = \beta_0 + \beta_1node + \epsilon$, where $\beta_1$ is the node count effect).

```{r 1_fit_alt_model, eval = FALSE}
# Fit the alternative regression model ----
alt_model1 <- fit_punc_model(
  data = data1,
  vcv_parts = vcv1,
  D = dmat1,
  model = "pn"  # path ~ node
)
```

```{r 1_fit_alt_model_save_object, eval = FALSE, echo = FALSE}
# To decrease knitting time, I save the created object above ----
save(alt_model1, file = "alt_model1.RData")
```

```{r 1_fit_alt_model_load_object, echo = FALSE}
# To decrease knitting time, I load the created object above ----
load("alt_model1.RData")  # alt_model1
```

```{r 1_show_alt_model_outputs}
# Print the model outputs ----
summary(alt_model1$model)
alt_model1$sig2  # variance
alt_model1$sse  # SSE
```

The y-intercept is the estimated mean path length when node count is zero ($0.543 \pm 0.044$ substitutions/site or $543 \pm 44$ substitutions/1,000 sites). But, this estimate is not in our interest here. We could center the node count so that the y-intercept is the value when the node count is at its mean. The slope or coefficient of the node count is $0.0040 \pm 0.0005$. For every additional net-speciation event, we see an increase in the estimated net molecular evolution by $4.0 \pm 0.5$ substitutions/1,000 sites. So, path length correlates with node count. And as expected, the calculated variability in path length after accounting for node count ($6.46$ substitutions/1,000 sites) is lower than the unconditional variance ($6.98$ substitutions/1,000 sites).

Below is the plot of the regression fit line.

```{r 1_plot_fit_line, fig.width = 4.75, fig.height = 2.94, fig.cap = "Fig. 3: Squamate molecular evolution seems punctuated. Higher node count (cumulative net-speciation events) corresponds with higher path lengths ($\\hat{\\beta_1} = 4.0 \\pm 0.5 \\text{ substitutions/1,000 sites}$). Regression fit line: $\\hat{path length} = 0.543 + 0.004 \\times node$"}
# Plot regression fit line ----
plot_punc_model(
  data = data1,
  output_reg = alt_model1,
  alpha = 0.15,
  type = "extant_molecular"
)
```

<br/>

The regression fit line is a bit off with respect to the data points because it accounts for the non-independence between path lengths (or the shared evolutionary history). Regardless of how the fit line looks, the data suggest non-linearity or a between-clade difference in the path length, which we should investigate in an actual study. The single line does not capture the pattern that, beyond a node count of ~25, a species may have a predicted path length of ~0.6 or ~1.1 substitutions per site, depending on which clade they belong to in the tree.

How significant is this result? The slope's $p\text{-value}$ is $<0.001$. There is a less than $0.001$ chance of observing a slope of $0.004$ or higher if we assume that the null is true (no correlation). However, following the American Statistical Association's (ASA's) statement on $p\text{-values}$, statistical significance does not equal practical significance^16^. The $p\text{-value}$ decreases as the sample size increases. I would argue that the following measures are more informative.

The $R^2$, calculated using the SST and the sum of squared estimates of error ($SSE$) from the null and alternative models, represents how much the explanatory variable explains the variation in the response.

```{r 1_calculate_r2}
# Calculate R-squared ----
r_squared1 <- calc_r2(output_null = null_model1, output_alt = alt_model1)
r_squared1 * 100
```

Node count explains $7.33%$ of the variability in path length. It means that $7.33%$ of the deviations from the strict molecular clock are attributable to net-speciation events. This value seems small, but it still falls within the 95% confidence interval reported by Pagel et al.^5^ for animals, plants, and fungi ($4.8%$ to $48%$).

The other statistic of interest is the proportion of the total amount of evolution attributable to punctuational effects, calculated using the formula $\frac{2(s-1)\beta}{T}$, where $s$ is the number of tips, $\beta$ is the slope, and $T$ is the sum of all branch lengths^5^.

```{r 1_estimate_punctuational_contribution}
# Estimate the punctuational contribution to the total divergence ----
punc_contrib1 <- est_punc_contrib(
  tree = tree_squamate_mol,
  output_reg = alt_model1
)
punc_contrib1 * 100
```

About $10%$ of the total squamate molecular evolution is attributable to the detected punctuational effect ($\frac{2(s-1)\hat{\beta}}{T} = 10.05\%$). This number is slightly lower than the lower bound of the 95% confidence interval reported by Pagel et al.^5^ for animals, plants, and fungi ($95\%\text{CI} = (10.6\%, 21.4\%)$).

To prevent overfitting, we compare the fit of the alternative model against that of the null model using the Bayesian Information Criterion ($BIC$^17^). A $BIC$ score difference higher than two ($\Delta BIC>2$) represents positive evidence for the model with the lowest BIC score, the best-fitting one^18^.

```{r 1_compare_models, eval = FALSE}
# Compare model fits with BIC
models1 <- list(null_model1, alt_model1)
compare_models(models = models1, file = NULL)
#> =========
#> BIC Table
#> =========
#>
#>         model       BIC delta_BIC
#> 1    path ~ 1 -19131.61     308.3
#> 2 path ~ node -19439.91       0.0
```

The regression model with node count as a predictor (alternative) has a much higher likelihood than the null model ($\Delta BIC=308.32$). We should keep node count as a predictor in the model.

However, note that the effect size, $R^2$, and the proportion of tree length attributable to the punctuational effect indicate that the detected punctuation pattern is weak to moderate at best. An increase in the predicted net molecular evolution of $4.0 \pm 0.5$ substitutions/1,000 sites for every additional net-speciation event may not be practically meaningful. If so, the question becomes why the punctuational effect is small in squamates broadly. Perhaps, the punctuation signal is undetectable when the phylogenetic tree is built using a dozen or so genes; a phylotranscriptomic or large-scale phylogenomic tree may be necessary. Potential statistical biases aside, the biological explanation likely lies in how speciation events typically occurred in Squamata. For example, does speciation usually happen by dispersal into new environments or passive vicariance (read Zink et al.^19^ for more discussion)? Kozak et al.^20^ argued that the rapid diversification of eastern North American woodland salamanders is non-adaptive, resulting from vicariant isolations that do not drive species to change their bioclimatic niche. So, the dispersal and vicariance frequencies among squamates may explain the low punctuational effect. Someone needs to estimate these frequencies. A good start would be a case study on the biogeography of Caribbean anole lizards^21^.

We need to check for potential statistical biases. A crucial one is the node-density artifact: branches tend to be shorter than their actual lengths in undersampled parts of a phylogenetic tree^22^. This artifact can bias the regression model towards detecting punctuated evolution even if this is not true (i.e., false positive). A symptom of this artifact would be a positive curvilinear relationship between the node count and path length ($\delta > 1$; see below).

```{r 1_test_node_density_artifact, eval = FALSE}
# Test for the node-density artifact ----
node_density1 <- est_node_density(data = data1, vcv_parts = vcv1)
```

```{r 1_test_node_density_artifact_save_object, eval = FALSE, echo = FALSE}
# To decrease knitting time, I save the created object above ----
save(node_density1, file = "node_density1.RData")
```

```{r 1_test_node_density_artifact_load_object, echo = FALSE}
# To decrease knitting time, I load the created object above ----
load("node_density1.RData")  # node_density1
```

```{r 1_show_node_density_model_outputs}
# Print the model outputs ----
summary(node_density1$model)
node_density1$delta
```

I fitted a log-log model above ($\widehat{ln(node)} = \beta_0 + \beta_1ln(path) = 2.38 + 0.82ln(path)$) and rearranged it so that the parameter interpretation is on the original scale ($\widehat{node} = \beta_0path^{\delta} = 10.80path^{0.82}$). There is little evidence for the node-density artifact since $\delta = 0.82 \text{ } \pm 0.08$. The fit line in the plot below does not seem curved.

```{r 1_plot_node_density_test, fig.width = 4.75, fig.height = 2.94, fig.cap = "Fig 4: The node-density artifact, which can bias the regression analysis, does not seem to be present ($\\delta = 0.82$). A symptom of this artifact, in this case, would be a positive curvilinear relationship ($\\delta > 1$). The fit line seems off because it arises from data points normalized against their evolutionary history (i.e., phylogenetically corrected)."}
# Plot the node-density test ----
plot_node_density(data = data1, output_nd = node_density1, unit = "DNA subs/site")
```

<br/>

We also still need to check the assumptions of a linear regression model^23,24^. Independence is an issue whenever we sample species, which is why our regression model accounts for the phylogenetic tree. Regarding the other assumptions, the plot below is a one-stop diagnostic shop.

```{r 1_plot_regression_diagnostics, fig.width = 4.75, fig.height = 2.94, fig.cap = "Diagnostics do not indicate violations of the normality and linearity assumptions. However, there is evidence of unequal variance as the variability of the phylogenetic residuals increases and then decreases with the fitted values."}
plot_diagnostics(output_reg = alt_model1)
```

<br/>

The distribution of the phylogenetic residuals is close to normal, and we see little-to-no changes in the average residuals across the fitted values (i.e., no indication of non-linearity). However, the constant variance assumption seems invalid. The violation of this assumption does not mean we should question the small punctuational effect we detected earlier, only the variance estimate.

[...RESUME: Simulate...]

[...Simulation: Compare the simulated vs. observed path length distribution...]

[...UPDATE: GitHub repo...]

## 2. Body size evolution in mammals

## 3. Genomic evolution in Zika viruses

## 4. Body size evolution in Mesozoic dinosaurs

## References

1. Darwin, C. *On the Origin of Species by Means of Natural Selection, or the Preservation of Favoured Races in the Struggle for Life*. (John Murray, 1859).

2. Mayr, E. Change of genetic environment and evolution. in *Evolution as a Process* (eds. Huxley, J., Hardy, A. C. & Ford, E. B.) 157–180 (Allan & Unwin, 1954).

3. Eldredge, N. & Gould, S. J. Punctuated equilibria: An alternative to phyletic gradualism. in *Models in Paleobiology* (ed. Schopf, T. J. M.) 82–115 (Freeman, Cooper, 1972).

4. Webster, A. J., Payne, R. J. H. & Pagel, M. Molecular phylogenies link rates of evolution and speciation. *Science* __301__, 478–478 (2003).

5. Pagel, M., Venditti, C. & Meade, A. Large punctuational contribution of speciation to evolutionary divergence at the molecular level. *Science* __314__, 119–121 (2006).

6. Atkinson, Q. D., Meade, A., Venditti, C., Greenhill, S. J. & Pagel, M. Languages evolve in punctuational bursts. *Science* __319__, 588–588 (2008).

7. Pagel, M., O'Donovan, C. & Meade, A. General statistical model shows that macroevolutionary patterns and processes are consistent with Darwinian gradualism. *Nat. Commun.* __13__, 1113 (2022).

8. Surya, K., Gardner, J. D. & Organ, C. L. Detecting punctuated evolution in SARS-CoV-2 over the first year of the pandemic. *Frontiers in Virology* __3__, (2023).

9. Pyron, R. A., Burbrink, F. T. & Wiens, J. J. A phylogeny and revised classification of Squamata, including 4161 species of lizards and snakes. *BMC Evol. Biol.* __13__, 93 (2013).

10. Tufte, E. R. *The Visual Display of Quantitative Information*. (Graphics Press, 1983).

11. Steel, E. A., Kennedy, M. C., Cunningham, P. G. & Stanovick, J. S. Applied statistics in ecology: Common pitfalls and simple solutions. *Ecosphere* __4__, art115 (2013).

12. Grafen, A. The phylogenetic regression. *Philos. Trans. R. Soc. Lond. B Biol. Sci.* __326__, 119–157 (1989).

13. Felsenstein, J. Phylogenies and the comparative method. *Am. Nat.* __125__, 1–15 (1985).

14. Symonds, M. R. E. & Blomberg, S. P. A primer on phylogenetic generalised least squares. in *Modern Phylogenetic Comparative Methods and Their Application in Evolutionary Biology: Concepts and Practice* (ed. Garamszegi, L. Z.) 105–130 (Springer Berlin Heidelberg, 2014).

15. Zuckerkandl, E. & Pauling, L. Evolutionary divergence and convergence in proteins. in *Evolving Genes and Proteins* (eds. Bryson, V. & Vogel, H. J.) 97–166 (Academic Press, 1965).

16. Wasserstein, R. L. & Lazar, N. A. The ASA statement on *p*-values: Context, process, and purpose. *Am. Stat.* __70__, 129–133 (2016).

17. Schwarz, G. Estimating the dimension of a model. *Ann. Stat.* __6__, 461–464 (1978).

18. Raftery, A. E. Bayesian model selection in social research. *Sociol. Methodol.* __25__, 111–163 (1995).

19. Zink, R. M., Blackwell-Rago, R. C. & Ronquist, F. The shifting roles of dispersal and vicariance in biogeography. *Proc. R. Soc. B* __267__, 497–503 (2000).

20. Kozak, K. H., Weisrock, D. W. & Larson, A. Rapid lineage accumulation in a non-adaptive radiation: Phylogenetic analysis of diversification rates in eastern North American woodland salamanders (Plethodontidae: *Plethodon*). *Proc. R. Soc. B* __273__, 539–546 (2006).

21. Losos, J. B. *Lizards in an Evolutionary Tree: Ecology and Adaptive Radiation of Anoles*. (Univ of California Press, 2011).

22. Venditti, C., Meade, A. & Pagel, M. Detecting the node-density artifact in phylogeny reconstruction. *Syst. Biol.* __55__, 637–643 (2006).

23. Freckleton, R. P. The seven deadly sins of comparative analysis. *J. Evol. Biol.* __22__, 1367–1375 (2009).

24. Mundry, R. Statistical Issues and Assumptions of Phylogenetic Generalized Least Squares. in *Modern Phylogenetic Comparative Methods and Their Application in Evolutionary Biology: Concepts and Practice* (ed. Garamszegi, L. Z.) 131–153 (Springer, 2014). doi:10.1007/978-3-662-43550-2_6.
